File: apps/fcc-extension/internal/protocol/spend_window.go
package protocol

import "math/big"

const MaxSpendWindowEntriesV1 = 4096

// SpendWindowEntryV1 is one ordered, reference-valued execution.
type SpendWindowEntryV1 struct {
	Value      *big.Int
	ExecutedAt uint64
}

// SpendWindowTotalsResultV1 contains the two deterministic cap aggregates.
type SpendWindowTotalsResultV1 struct {
	DailySpend   *big.Int
	RollingSpend *big.Int
}

// SpendWindowTotalsV1 sums UTC calendar-day and exact sliding-window history.
func SpendWindowTotalsV1(entries []SpendWindowEntryV1, now, rollingWindowSeconds uint64) (SpendWindowTotalsResultV1, bool) {
	if rollingWindowSeconds == 0 || len(entries) > MaxSpendWindowEntriesV1 {
		return SpendWindowTotalsResultV1{}, false
	}
	dayStart := now / 86400 * 86400
	hasLowerBound := now >= rollingWindowSeconds
	rollingLowerBound := uint64(0)
	if hasLowerBound {
		rollingLowerBound = now - rollingWindowSeconds
	}
	dailySpend, rollingSpend := new(big.Int), new(big.Int)
	var priorTimestamp uint64
	for index, entry := range entries {
		if entry.Value == nil || entry.Value.Sign() <= 0 || entry.Value.Cmp(maxUint256) > 0 || entry.ExecutedAt > now || (index > 0 && entry.ExecutedAt < priorTimestamp) {
			return SpendWindowTotalsResultV1{}, false
		}
		priorTimestamp = entry.ExecutedAt
		if entry.ExecutedAt >= dayStart {
			if entry.Value.Cmp(new(big.Int).Sub(new(big.Int).Set(maxUint256), dailySpend)) > 0 {
				return SpendWindowTotalsResultV1{}, false
			}
			dailySpend.Add(dailySpend, entry.Value)
		}
		if !hasLowerBound || entry.ExecutedAt > rollingLowerBound {
			if entry.Value.Cmp(new(big.Int).Sub(new(big.Int).Set(maxUint256), rollingSpend)) > 0 {
				return SpendWindowTotalsResultV1{}, false
			}
			rollingSpend.Add(rollingSpend, entry.Value)
		}
	}
	return SpendWindowTotalsResultV1{DailySpend: dailySpend, RollingSpend: rollingSpend}, true
}

File: packages/protocol/src/spend-window.ts
const MAX_UINT256 = (1n << 256n) - 1n;
const MAX_UINT64 = (1n << 64n) - 1n;
export const MAX_SPEND_WINDOW_ENTRIES_V1 = 4_096;

export interface SpendWindowEntryV1 {
  value: bigint;
  executedAt: bigint;
}

export interface SpendWindowTotalsV1 {
  dailySpend: bigint;
  rollingSpend: bigint;
}

/** Sums a UTC calendar day and an exact sliding window from ordered history. */
export function spendWindowTotalsV1(
  entries: readonly SpendWindowEntryV1[],
  now: bigint,
  rollingWindowSeconds: bigint,
): SpendWindowTotalsV1 | null {
  if (now < 0n || now > MAX_UINT64 || rollingWindowSeconds <= 0n
    || rollingWindowSeconds > MAX_UINT64 || entries.length > MAX_SPEND_WINDOW_ENTRIES_V1) return null;
  const dayStart = (now / 86_400n) * 86_400n;
  const hasLowerBound = now >= rollingWindowSeconds;
  const rollingLowerBound = hasLowerBound ? now - rollingWindowSeconds : 0n;
  let dailySpend = 0n;
  let rollingSpend = 0n;
  let priorTimestamp = 0n;
  for (let index = 0; index < entries.length; index += 1) {
    const entry = entries[index]!;
    if (entry.value <= 0n || entry.value > MAX_UINT256 || entry.executedAt < 0n
      || entry.executedAt > now || entry.executedAt > MAX_UINT64
      || (index > 0 && entry.executedAt < priorTimestamp)) return null;
    priorTimestamp = entry.executedAt;
    if (entry.executedAt >= dayStart) {
      if (entry.value > MAX_UINT256 - dailySpend) return null;
      dailySpend += entry.value;
    }
    if (!hasLowerBound || entry.executedAt > rollingLowerBound) {
      if (entry.value > MAX_UINT256 - rollingSpend) return null;
      rollingSpend += entry.value;
    }
  }
  return { dailySpend, rollingSpend };
}
